---
# Problem 1:Market-share

---
Suppose tha in a country, three major companies provide high-speed internet : Companies A, B and  C.

Most of the users subscribe to one of these companies (the market share of local companies is negligible). The percentages of subscibers to each company is decribed in the table below


Company: |A | B |C |
---|---|---| ---|
% of market share: |0.35 | 0.15 | 0.50

Each year, a portion of customers of one company decides to switch to a different company.

By the end of 2019 :
- Company A  lost 30% of its customers to B and 40% of its customers to C. (Therefore A retained 30% of its own customers.)
- Company B lost 20% of its customers to A and 10% of its customers to C. (Therefore B retained 70% of its own customers.)
- Company  C lost 25% of its customers to A, and 20% of its customers to B. (Therefore C retained 55% of its own customers.)


We can summarize the above data in the following table :

from | A | B | C | to
---|---|---|---|---|
 | 0.3 | 0.2 | 0.25 | A
  | 0.3 | 0.7 | 0.2 | B
  | 0.4 | 0.1 | 0.55 | C


 The colums of the table represent the proportion of subscribers that swtiched  from one company to another. For example, the second column refers to costumers form company B. Reading from the top down we see that  that 20% of the costumers from B moved to A, 70% of the costumers stayed in B, and 10 % of the costumers moved to company C.


  As a matrix this can be expressed as

$M=\begin{pmatrix}
0.3 & 0.2& 0.25\\
0.3& 0.7 & 0.2 \\
0.4 & 0.1 & 0.55
\end{pmatrix}$


**Note** ths sum of each column equals to 1.00., why?

---

## Question

What would be the market share of each company after 1 year?


If we call $A_1$, $B_1$ , $C_1$ the market shares after 1 year,
Then we have the eqs

$A_1= 0.3A_0+.2 B_0 +.25C_0\\
B_1= 0.3A_0 + 0.7B_0+ 0.2C_0\\
C_1=0.4A_0+0.1B_0+0.55C_0
$


This can be seen as a Matrix-vector product:

$\begin{pmatrix}
A_1\\B_1\\C_1
\end{pmatrix}=
\begin{pmatrix}
0.3 & 0.2& 0.25\\
0.3& 0.7 & 0.2 \\
0.4 & 0.1 & 0.55
\end{pmatrix}
\begin{pmatrix}
A_0\\B_0\\C_0
\end{pmatrix}$


We do this calculation below:

In [1]:
from sympy import *
M=Matrix([[.3, .2, .25], [.3, .7, .2], [.4, .1, .55] ] )
v=Matrix([[.35], [.15], [.5]])
M*v

Matrix([
[0.26],
[0.31],
[0.43]])

### Problem **1A**
What will be the market share of each company after two years?

In [2]:
# Write the code for Problem 1 here
M*M*v

Matrix([
[0.2475],
[ 0.381],
[0.3715]])

### Problem **1B**
What will be the market share of each company after 8 years?

In [3]:
# write the code for answering problem 2 here
M**8 *v

Matrix([
[0.239687019023437],
[0.447007135734375],
[0.313305845242188]])

---

## Terminology

_**Defnition**_ A vector $\mathbf{v}$ is called a **probability vector** if
- its entries are nonnegative (they *could be zero))
- the sum of the entries are equal to 1.


- A square matrix $P$ is called a **stochastic matrix** if its colums are *probability vectors*

### Examples

- $\mathbf{v}= \begin{pmatrix} 0.3\\ 0.3 \\0.2 \\ 0.2 \end{pmatrix}$ is a probability vector.


- the matrix $P= \begin{pmatrix} 0.3 & 0.4 & 0.24\\ 0.3 & 0.6 & 0.16 \\0.1 & 0.0 & 0.6 \end{pmatrix}$ is a Stochastic matrix.

### Markov Chains

A **Markov Chain** is a sequence of probability vectors:

$\left\{ \mathbf{v_0}, \mathbf{v_1}, \ldots \mathbf{v_k}, \mathbf{v_{k+1} \ldots}  \right\}$

together with a matrix $M$, such that

$\mathbf{v_{k+1}}=M\mathbf{v_k}$

- The vectors $\mathbf{v_k}$ are called state vectors,
- The matrix is called the **migration matrix**, or **transition matrix**

*italicized text* ### Regular Stochastic Matrix
  A stochastic matrix $P$ is called **regular** if for some $k \geq 1$, all of the entries of $P^k$ are strictly positive.

 ### Steady-state vector
If $M$ is an $n \times n$ stochastic matrix, then a nonzero vector $\mathbf{v} \in R^n$ is called a **steady-state** vector or an **equilibrium** vector for $M$ if it has the following two properties :
- $\mathbf{v}$ is a probability vector, i.e., its entries are non-negative and the sum of entries of $\mathbf{v}$ equals 1.
- $M\mathbf{v} = \mathbf{v}$, i.e., $\mathbf{v}$ is an eigenvector of M corresponding to eigenvalue λ = 1

### Existence of Equilibrium (Steady-state)  vectors

_**Theorem**_

If $M$ is a regular stochastic matrix , then it has a *unique* stady-state vector $\mathbf{v}$.

Moreover, if

$\left\{ \mathbf{v_0}, \mathbf{v_1}, \ldots \mathbf{v_k}, \mathbf{v_{k+1} \ldots}  \right\}$

is a Markov Chain with initial state $\mathbf{v_0}$, and transition matrix $M$, then

$\displaystyle \lim_{n\to \infty }\mathbf{v_i} = \mathbf{v}$.

 This is equivallent to

$\displaystyle \lim_{n \to \infty} M^n ( \mathbf{v_0})= \mathbf{v}$


If the migration matrix, $M$,  of a Markov chain is regular , then
- the long term behavior of the Markov chain is decribed by the steady-state vector
- The eigenspace of $M$ corresponsding to the eigenvalue $\lambda = 1$ is one-dimensional
- The steady-state  vector is the unique probability vector in the eigenspace $E_1$, corresponding to the eigenvalue $\lambda=1$
- the long-term behavior does not depend on the initial state.


### Problem 1C
   In the long run, what will be the market share of each of the companies?

In [4]:
M**9999999 * v

Matrix([
[ 0.23958333332817],
[0.447916666657014],
[0.312499999993266]])

### Problem 1D

Find the Eigenvectors and eigenvalues for the matrix $M$
(Add below as many liness of codes as you need)


In [5]:
M_eigenvects = M.eigenvects()
M_eigenvects

[(0.0613999063670617,
  1,
  [Matrix([
   [-0.778978272455984],
   [ 0.177605587359474],
   [  0.60137268509651]])]),
 (0.488600093632938,
  1,
  [Matrix([
   [-0.0867466272543137],
   [  0.760941576721673],
   [ -0.674194949467359]])]),
 (1.00000000000000,
  1,
  [Matrix([
   [0.414970505980044],
   [ 0.77581442422356],
   [0.541265877365274]])])]

## Problem 1E

Find the steady-state vector by finding a probability vector $\mathbf{v} \in E_1$.
Compare this with your answer to problem 1C

In [6]:
eigenvector = M_eigenvects[2][2][0]
# find divisor so that sum = 1
x = 1/(eigenvector[0] + eigenvector[1] +eigenvector[2])
steady_state_vector = eigenvector * x
# check
steady_state_vector[0] + steady_state_vector[1] + steady_state_vector[2]

1.00000000000000

In [7]:
steady_state_vector

Matrix([
[0.239583333333333],
[0.447916666666667],
[           0.3125]])

In [8]:
# check
M * steady_state_vector

Matrix([
[0.239583333333333],
[0.447916666666667],
[           0.3125]])

Steady State Vector = \begin{pmatrix}
  .239583333333333\\
0.447916666666667\\
0.3125\\
\end{pmatrix}

---

#Problem 2: Votes in elections
---
 Suppose that a country has 3 political parties: the Conservative (C), Liberal (L), and National (N) parties. If a person votes for the candidate from one party in an election, that person may decide to vote for the same party in the next election or may switch to vote for a candidate from another party in the next election.

The percentages are represented in the table

from | C | L |N | to
---|---|---|---|---|
 | 0.5 | 0.4 | 0.10 | C
  | 0.3 | 0.4 | 0.3 | L
  | 0.2 | 0.2 | 0.6 | N



### Problem 2A
If in this election Conservatives received 25% of the votes, Liberals 30% of the votes, and Nationals the remaining 45% of the votes, what is the predicted distribution for the next election?



In [9]:
M=Matrix([[.5, .4, .1], [.3, .4, .3], [.2, .2, .6] ] )
M

Matrix([
[0.5, 0.4, 0.1],
[0.3, 0.4, 0.3],
[0.2, 0.2, 0.6]])

In [10]:
v=Matrix([[.25], [.30], [.45]])
v

Matrix([
[0.25],
[ 0.3],
[0.45]])

In [11]:
M*v

Matrix([
[0.29],
[0.33],
[0.38]])

### Problem 2B
If a person voted for the National party in this election, find the probability that the person votes for the Conservative party in the election two years from now.**bold text**

In [12]:
M**2

Matrix([
[0.39, 0.38, 0.23],
[0.33, 0.34, 0.33],
[0.28, 0.28, 0.44]])

**.23** is the probability that the person votes for the Conservative party two years from now.

### Problem 2C
If this transition matrix is valid for many years, what is going to happen with the distibution of votes in the long term?

In [13]:
M**9999999 *v

Matrix([
[0.333333333391551],
[0.333333333391552],
[0.333333333391551]])

Long term the votes will be very close to evenly distributed.

### Problem 2D

Use eigenvectors to **find** the steady-state vector for this problem.

In [14]:
M.eigenvects()

[(0.100000000000000,
  1,
  [Matrix([
   [    0.707106781186548],
   [   -0.707106781186547],
   [-2.90757862719218e-17]])]),
 (1.00000000000000,
  1,
  [Matrix([
   [0.577350269189626],
   [0.577350269189626],
   [0.577350269189626]])]),
 (0.400000000000000,
  1,
  [Matrix([
   [    0.816496580927726],
   [-3.77705543197366e-17],
   [   -0.816496580927726]])])]

In [15]:
eigenvector = M.eigenvects()[1][2][0]
# find divisor so that sum = 1
x = 1/(eigenvector[0] + eigenvector[1] +eigenvector[2])
steady_state_vector = eigenvector * x
# check
steady_state_vector[0] + steady_state_vector[1] + steady_state_vector[2]

1.00000000000000

In [16]:
steady_state_vector

Matrix([
[0.333333333333333],
[0.333333333333333],
[0.333333333333333]])

---
## Problem 3: Website search
---

Markov chains play an important role in online search.

“PageRank is an algorithm used by Google Search to rank websites in their search engine results. PageRank was named after Larry Page, one of the founders of Google. PageRank is a way of measuring the importance of website pages”
Source: https://en.Wikipedia.org/wiki/PageRank under the Creative Commons Attribution-ShareAlike License;

The theory behind PageRank is that pages that are linked to more often are more important and useful; identifying those that are linked to more often about a topic helps identify the pages that should be presented as most pertinent in a search.

In real world search, there are thousands or millions of pages linking together, resulting in huge transition matrices. Because of the size and other properties of these matrices, the mathematics behind PageRank is more sophisticated than the small example we examine here with only four websites. However our example is adequate to convey the main concept of PageRank and its use in search algorithms.

It should be noted that real world search algorithms, PageRank or similar Markov chain ranking schemes are only one of a variety of methodsused.

Suppose we have 4 webpages that contains links to each other. We call the pages A, B, C, D.

- From page A, 30% of people link to page B, 50% link to page C, and 20% link to page D
- From page B, 50% of popele link to page A and 50% link to page D
- From page C, 10% of people link to page B, 70% link to page C, and 20% link to page D
- From page D, 20% of people link to page A, 40% to page B, 10% to page C, and 30% link to page D

(In this example, when a page links to itself, it means that a person viewing the page stays at that page and does not link to another page.)

### Problem 3A
Write the transition matrix.







In [17]:
M = Matrix([
    [0, .5, 0, .2],
    [.3, 0, .1, .4],
    [.5, 0, .7, .1],
    [.2, .5, .2, .3]])
M

Matrix([
[  0, 0.5,   0, 0.2],
[0.3,   0, 0.1, 0.4],
[0.5,   0, 0.7, 0.1],
[0.2, 0.5, 0.2, 0.3]])


### Problem 3B
Is this a regular Markov chain? Explain


In [18]:
M**2

Matrix([
[0.19,  0.1, 0.09, 0.26],
[0.13, 0.35, 0.15, 0.19],
[0.37,  0.3, 0.51,  0.2],
[0.31, 0.25, 0.25, 0.35]])

Yes because there are values of $M^n$ that don't contain any zeros like $M^2$ shown above.

### Problem 3C
Find the equilibrium vector and write a sentence summarizing the long term distribution of visits to these sites based on this model.



In [19]:
M.eigenvects()

[(1.00000000000000,
  1,
  [Matrix([
   [-0.299051620152331],
   [-0.378001247872547],
   [-0.681837693947316],
   [ -0.55025498108029]])]),
 (0.452132154830300,
  1,
  [Matrix([
   [-0.395189281909232],
   [-0.266527884660268],
   [ 0.888786362896129],
   [-0.227069196326629]])]),
 (-0.442127147315265,
  1,
  [Matrix([
   [ 0.632301492662817],
   [-0.718057557228767],
   [-0.311599552534991],
   [ 0.397355617100941]])]),
 (-0.0100050075150345,
  1,
  [Matrix([
   [ 0.974246405830471],
   [ 0.214558740064842],
   [-0.603671582674063],
   [ -0.58513356322125]])])]

In [20]:
eigenvector = M.eigenvects()[0][2][0]
# find divisor so that sum = 1
x = 1/(eigenvector[0] + eigenvector[1] + eigenvector[2] + eigenvector[3])
x
steady_state_vector = eigenvector * x
steady_state_vector

Matrix([
[0.156641604010025],
[0.197994987468672],
[0.357142857142857],
[0.288220551378446]])

### Problem 3D
In the equilibrium vector, the state with the highest probability has the highest “Page-rank” and as the probabilities decrease, the ranking decreases. Indicate the order of the ranking, from highest Page rank to lowest Page rank, of these 4 pages.


In [21]:
M * steady_state_vector

Matrix([
[0.156641604010025],
[0.197994987468672],
[0.357142857142857],
[0.288220551378446]])

1. C
2. D
3. B
4. A